heron/packing/src/java/com/twitter/heron/packing/builder/HomogeneityScorer.java

```diff
@@ -0,0 +1,51 @@
+// Copyright 2016 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.twitter.heron.packing.builder;
+
+import com.twitter.heron.spi.packing.PackingPlan;
+
+/**
+ * Scores based on how homogeneous a container is relative to a given component, from highest to
+ * lowest. So a container with all the same components (e.g., AAAA) will get the highest score for
+ * A, while a container with a mix of many components (e.g. ABCD) will get a low score.
+ */
+public class HomogeneityScorer implements Scorer<Container> {
+  private String componentName;
+
+  public HomogeneityScorer(String componentName) {
+    this.componentName = componentName;
+  }
+
+  @Override
+  public boolean sortAscending() {
+    return false;
+  }
+
+  @Override
+  public double getScore(Container container) {
+    int totalComponentInstances = 0;
+    int totalInstances = container.getInstances().size();
+
+    if (totalInstances == 0) {
+      return 0;
+    }
+
+    for (PackingPlan.InstancePlan instancePlan : container.getInstances()) {
+      if (componentName.equals(instancePlan.getComponentName())) {
+        totalComponentInstances++;
+      }
+    }
+    return (double) totalComponentInstances / totalInstances;
```

Comment:
As implemented we're trying to achieve balance while also removing containers where we can (despite how I named this PR). So I would argue that we would want to go from [BBCD, ABC] to [BCD, ABC] which is a more balanced state. To expand on this example, if we instead had [BBBBBBCD, ABC], we should try to removing as many Bs as possible from the first container to reduce our imbalance before trying to remove the seconds container.